text/0006-monolothic-packaging.md

```diff
@@ -0,0 +1,343 @@
+JDK, the .NET Framework. When users write Node.js libraries or apps they don't "depend" on the Node.js SDK, they actually "peer depend" on it (it is expressed via a special `engines` section in `package.json` but for all intents and purposes it's behaves like a peer dependency).
+
+I would argue that if a vendor publishes a 3rd-party construct library, what they *really* want to say is "I am compatible with CDK >= 1.23.0". Then, the decision about which actual CDK version is being used is left to the app level.
+
+## Alternatives Considered
+
+### Tooling
+
+TODO: we considered building additional tooling (e.g. `cdk install`) that will make hyper-modular peer dependencies sane.
+
+### Meta-package
+
+This approach proposes that we keep the hyper modularity but also ship a meta-package that will either just take a dependency on all modules (like `decdk` is doing today) or also "re-export" all the types.
+
+The main benefit is that it will technically allow the interoperability of the
+two modes. Libraries can still take granular peer dependencies while apps will
+depend on the meta package.
+
+TODO: we need to check if in JavaScript it is possible to avoid the re-export and only create a module that brings in all the other modules as dependencies. The answer is probably "depends on the package manager". Yarn pre-2.0 for example would hoist all the transitive deps so it would "just work", but post 2.0 they actually disallow that.
+
+# Adoption Strategy
+
+TODO
+
+# Unresolved questions
+
+TODO
+
+# Future Possibilities
+
+TODO
+
+# Appendix
+
+## Peer Dependencies in npm
+
+> Based on a whitepaper by [@rix0rrr](https://github.com/rix0rrr).
+
+Most languages (Java, Python, .NET) only support a single instance of a module
+at runtime. This means that if two modules declare that they take a dependency
+on a third module, the dependency manager will resolve a common version and
+fetch it.
+
+Contrary to most languages, the Node.js ecosystem supports the co-existence of
+multiple versions of a module at runtime (since `node_modules` first searched as
+a relative path). This is problematic for the CDK case since CDK modules often
+expose types from other modules in their API (e.g. an `s3.Bucket` accepts a
+`kms.Key` as an option).
+
+The ultimate root cause of the issues we are encountering is a specific feature of the NPM package manager, one that both encourages a package ecosystem that “just works” with minimal maintenance and versioning problems, as well one that significantly simplifies the package manager:
+
+_Packages can appear multiple times in the dependency graph, at different versions._
+
+It simplifies the package manager because the package manager does not have to resolve version conflicts to arrive at a single version that will work across the entire dependency tree, and report a usable error if fails to do so. In the presence of version ranges, this is an open research problem, which NPM conveniently sidesteps.
```

Comment:
```suggestion
It simplifies the package manager because the package manager does not have to resolve version conflicts to arrive at a single version that will work across the entire dependency tree, and report a usable error if it fails to do so. In the presence of version ranges, this is an open research problem, which NPM conveniently sidesteps.
```